### PR TITLE
Port lu_solve to ATen

### PR DESCRIPTION
**Pull request by vishwakftw**
Changelog:
- Port TH implementation to ATen/native/BatchLinearAlgebra.cpp
- Port THC implementation to ATen/native/cuda/BatchLinearAlgebra.cu
- Remove TH/THC implementations
- Update doc strings

Test Plan:
- Existing tests should pass to confirm that the port is successful

### CODE REVIEW

**Comment by t-vi** [comment]
Maybe I missed it, but woudln't you want to check the device of the (non-self) arguments somewhere?

**Comment by vishwakftw** [comment]
`_linear_solve_broadcast_args` defined in LinearAlgebraUtils.h ensures that this is not the case. However, I haven't checked the device for the pivots tensor (thanks for reminding), and I'll add that check for it right away.

**Review by t-vi — APPROVED** [review]
Looks good to me.

**Comment by vishwakftw** [comment]
Thanks for the review.

**Comment by zhangguanheng66** [comment]
@zou3519 Do you mind if you could take a quick look before landing the PR? Thanks.

**Comment by vishwakftw** [comment]
Just a quick note @zou3519, this PR is bc-breaking in the following way: the b needs to have  trailing dimension for the number of the right hand sides. This means that b.lu_solve(a, p) where b is 3 x 2 and a is 3 x 2 x 2 will throw an error. Instead users will have to unsqueeze the last dimension of b to get 3 x 2 x 1. 

Motivation behind this change: ensure consistency with all solve methods (solve, cholesky_solve, triangular_solve) which also require the trailing dimension.

**Comment by zou3519** [comment]
@vishwakftw thanks for the clarification. 

How does the error message look like if one does `b.lu_solve(a, p)` where b is 3 x 2 and a is 3 x 2 x 2? It might be better if we deprecate the old behavior first (by adding its functionality and a warning) and then remove it completely in the future, but a good error message might be sufficient here.

**Comment by vishwakftw** [comment]
The error message could generally pop up at two points in the current version (i.e., this PR)
- inferring sizes of the broadcasted operands.
- mismatch of `n` in `b` i.e. `lda` and `n` in `a`

In your use case, the source of the error would be the 2nd case. This is because b with size 3 x 2 has no batch dimensions, while a with size 3 x 2 x 2 has batch dimensions (3,). This would mean that b is expanded to 3 x 3 x 2. It is now easy to see that `A @ x = b` cannot be mathematically possible, therefore leading to a mismatch.

I think the previous way of doing things can be brought back with an conditional expand. What I mean here is that we first check if broadcasting can be done, if not unsqueeze the last dimension and then broadcast. This could be prone to a multitude of unintentional errors.

**Comment by vishwakftw** [comment]
@zou3519 is this good to go?

**Comment by zou3519** [comment]
Sorry, I haven't had time to think about how BC-breaking this change actually is, I will do that later today.

**Comment by zou3519** [comment]
I think we should definitely do a deprecation cycle of some sort because modifying the API like this could break user code. 

> I think the previous way of doing things can be brought back with an conditional expand. What I mean here is that we first check if broadcasting can be done, if not unsqueeze the last dimension and then broadcast. This could be prone to a multitude of unintentional errors.

Did you add the broadcasting batch dimensions behavior in this PR or did it exist already?

**Comment by vishwakftw** [comment]
> Did you add the broadcasting batch dimensions behavior in this PR or did it exist already?

The broadcasting batch dimensions was added in this PR. The earlier version of lu_solve only accepted 3D tensors.

**Comment by zou3519** [comment]
> > Did you add the broadcasting batch dimensions behavior in this PR or did it exist already?
> 
> The broadcasting batch dimensions was added in this PR. The earlier version of lu_solve only accepted 3D tensors.

Okay, so I think we have a couple of options here:
1. Deprecate the A: (3, 2, 2) and B: (3, 2) behavior in the next release, then add broadcasting of batch dimensions (this change). This makes the timeline for the feature kind of slow.
2. We could add a flag that does the legacy behavior by default and when switched on does broadcasting of batch dimensions and say the legacy behavior is deprecated. Removing the flag in the future might be a little hard because of how TorchScript works.
3. Do nothing, land this PR as-is, and hope for the best

**Comment by vishwakftw** [comment]
I think 1 is the better option - in view of usability and UX. I'll modify the PR accordingly.

**Comment by zou3519** [comment]
> I think 1 is the better option - in view of usability and UX. I'll modify the PR accordingly.

Thank you, let's go for that then

**Comment by vishwakftw** [comment]
@zou3519 I've reverted the broadcasting behavior, and the port is a direct conversion to ATen from TH/THC (sans the backend change from cuBLAS to MAGMA).

**Review by zou3519 — COMMENTED** [review]
I'm still reading through the PR, but because you did a port from cublas to magma some before/after numbers would be great to have to make sure it didn't get slower.

File: torch/_torch_docs.py [hunk 1]
```diff
@@ -2653,24 +2653,23 @@ def merge_dicts(*dicts):
...
-Batch LU solve.
-
 Returns the LU solve of the linear system :math:`Ax = b` using the partially pivoted
 LU factorization of A from :meth:`torch.lu`.
 
 Arguments:
-    b (Tensor): the RHS tensor
-    LU_data (Tensor): the pivoted LU factorization of A from :meth:`torch.lu`.
-    LU_pivots (IntTensor): the pivots of the LU factorization
+    b (Tensor): the RHS tensor of size :math:`(b, m, *)`, where :math:`*` is a zero or one
```
Comment (zou3519) [inline torch/_torch_docs.py]: Maybe we shouldn't document that `*` can be zero dimensions to discourage use of it.

File: test/test_torch.py [hunk 2]
```diff
@@ -5420,6 +5441,18 @@ def test_solve_methods_arg_device(self):
...
+            A = A.unsqueeze(0)
+            with self.assertRaisesRegex(RuntimeError, err_str):
+                torch.lu_solve(b, A, torch.rand(A.shape[:-1], device=A_device).int())
+
+            # This checks if a suitable error message is thrown
+            # when LU output and pivots are on the same device
+            with self.assertRaisesRegex(RuntimeError,
+                                        "Expected LU_pivots and LU_data to be on the same device"):
+                torch.lu_solve(b, A, torch.rand(A.shape[:-1], device=b_device).int())
+
```
Comment (zou3519) [inline test/test_torch.py]: Can we add an extra test for the deprecation warning?

File: aten/src/ATen/native/BatchLinearAlgebra.cpp [hunk 3]
```diff
@@ -1066,4 +1084,97 @@ std::tuple<Tensor&, Tensor&, Tensor&> svd_out(Tensor& U, Tensor& S, Tensor& VT,
...
+  });
+  if (self.dim() > 2) {
+    batchCheckErrors(infos, "lu_solve_cpu");
+  } else {
+    singleCheckErrors(infos[0], "lu_solve_cpu");
+  }
+  return self_working_copy;
+}
+
+// Supports arbitrary batch dimensions for self and LU_data (implicity LU_pivots also)
```
Comment (zou3519) [inline aten/src/ATen/native/BatchLinearAlgebra.cpp]: This comment is outdated now, right?

File: aten/src/ATen/native/cuda/BatchLinearAlgebra.cu [hunk 4]
```diff
@@ -1371,6 +1417,93 @@ std::tuple<Tensor, Tensor, Tensor> _svd_helper_cuda(const Tensor& self, bool som
...
+    auto b_stride = matrixStride(b);
+    auto lu_stride = matrixStride(lu);
+    auto pivots_stride = pivots.size(-1);
+    magma_int_t batch_size = magma_int_cast(batchCount(b), "batchCount");
+
+    magma_int_t** pivots_array;
+    scalar_t** lu_array;
+    scalar_t** b_array;
+
+    ALLOCATE_ARRAY(pivots_array, magma_int_t*, batch_size, b);
```
Comment (zou3519) [inline aten/src/ATen/native/cuda/BatchLinearAlgebra.cu]: This needs a rebase, ALLOCATE_ARRAY was updated to use 3 args

**Review by zou3519 — COMMENTED** [review]
Code looks good. Would be good to see some performance numbers for the CUDA version since you switched it to magma. Also added some minor comments elsewhere:
- needs a rebase
- we should add a test for the deprecation warning

File: aten/src/ATen/native/BatchLinearAlgebra.cpp (hunk 3 above)
Comment (vishwakftw) [reply aten/src/ATen/native/BatchLinearAlgebra.cpp]: Sorry, forgot to clean it up. I’ll do it.

File: aten/src/ATen/native/cuda/BatchLinearAlgebra.cu (hunk 4 above)
Comment (vishwakftw) [reply aten/src/ATen/native/cuda/BatchLinearAlgebra.cu]: Sure, this should be fixed in the rebase.

File: test/test_torch.py (hunk 2 above)
Comment (vishwakftw) [reply test/test_torch.py]: I’ll add this as well.

File: torch/_torch_docs.py (hunk 1 above)
Comment (vishwakftw) [reply torch/_torch_docs.py]: I think this seems reasonable. Will modify accordingly.

**Comment by vishwakftw** [comment]
Thank you for the reviews @zou3519, I’ll make the changes as suggested and have some benchmark numbers by next afternoon.

**Comment by vishwakftw** [comment]
@zou3519 These are the benchmark results (couldn't run really large results locally due to 2GB limit): https://gist.github.com/vishwakftw/94a985561a88cc2ecef9b1473e4cbaa0

TL;DR: MAGMA is slower on smaller matrix sizes than cuBLAS but performs way better on larger matrix sizes. However, for almost all the tested sizes, running lu_solve on CPU and moving to CUDA is the fastest option.

**Review by zou3519 — APPROVED** [review]

File: aten/src/ATen/native/BatchLinearAlgebra.cpp [hunk 5]
```diff
@@ -196,6 +206,14 @@ template<> void lapackSvd<float>(char jobz, int m, int n, float *a, int lda,
                                  float *s, float *u, int ldu, float *vt, int ldvt, float *work, int lwork, int *iwork, int *info) {
   sgesdd_(&jobz, &m, &n, a, &lda, s, u, &ldu, vt, &ldvt, work, &lwork, iwork, info);
 }
+
+template<> void lapackLuSolve<double>(int n, int nrhs, double *a, int lda, int *ipiv, double *b, int ldb, int *info) {
+  dgetrs_("N", &n, &nrhs, a, &lda, ipiv, b, &ldb, info);
```
Comment (zou3519) [inline aten/src/ATen/native/BatchLinearAlgebra.cpp]: Is this correct? Shouldn't it be `'N'` (the character version?) I'm not sure why this works in string form

File: aten/src/ATen/native/BatchLinearAlgebra.cpp [hunk 6]
```diff
@@ -196,6 +206,14 @@ template<> void lapackSvd<float>(char jobz, int m, int n, float *a, int lda,
...
+}
+
+template<> void lapackLuSolve<float>(int n, int nrhs, float *a, int lda, int *ipiv, float *b, int ldb, int *info) {
+  sgetrs_("N", &n, &nrhs, a, &lda, ipiv, b, &ldb, info);
```
Comment (zou3519) [inline aten/src/ATen/native/BatchLinearAlgebra.cpp]: Ditto

File: aten/src/ATen/native/BatchLinearAlgebra.cpp (hunk 5 above)
Comment (vishwakftw) [reply aten/src/ATen/native/BatchLinearAlgebra.cpp]: Oh yes, I’m sorry. I’ll fix this right away.

Comment (zou3519) [reply aten/src/ATen/native/BatchLinearAlgebra.cpp]: No worries, I'm just really surprised that our builds allow this conversion from string literal to char

File: aten/src/ATen/native/BatchLinearAlgebra.cpp [hunk 7]
```diff
@@ -56,6 +57,10 @@ extern "C" void dgesdd_(char *jobz, int *m, int *n, double *a, int *lda,
                         double *s, double *u, int *ldu, double *vt, int *ldvt, double *work, int *lwork, int *iwork, int *info);
 extern "C" void sgesdd_(char *jobz, int *m, int *n, float *a, int *lda,
                         float *s, float *u, int *ldu, float *vt, int *ldvt, float *work, int *lwork, int *iwork, int *info);
+
+// getrs
+extern "C" void dgetrs_(char *trans, int *n, int *nrhs, double *a, int *lda, int *ipiv, double *b, int *ldb, int *info);
```
Comment (zou3519) [inline aten/src/ATen/native/BatchLinearAlgebra.cpp]: Do you know where our lapack is coming from? From reading random online sources (https://software.intel.com/en-us/mkl-developer-reference-c-getrs) I think the first arg should be a `char`, but maybe it's actually supposed to be a `char *` (and take a string literal) depending on what we're linking with

Comment (t-vi) [reply aten/src/ATen/native/BatchLinearAlgebra.cpp]: Note that those are LAPACKE functions (with an extra `int matrix_layout`, too), openblas has the function as declared here:
https://github.com/xianyi/OpenBLAS/blob/6b41eb9c0cf1ce8c8a10a8e7fd3a6f8cbd94a14a/common_interface.h#L675
I can't find it in the online doc for MKL, but the dgetrs declaration in my installed `mkl_lapack.h` is also with `char*`.
(Reminds me of the only tangentially related https://lwn.net/Articles/791393/, but that's not of worry now.)

**Comment by vishwakftw** [comment]
@zou3519 Instead of passing a character, I passed a reference to it, and it seems to have solved the issue. This was done in the btrisolve implementation that existed in TH.
